Support DynamicSlicing in torchdynamo

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #90410

Currently, i am limiting the step size to be normal integer, just because I am not really sure what the expected behaviour for x[None:None:symint] as the original python behaviour adjusts start and stop based on step. 


cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

Differential Revision: [D42049013](https://our.internmc.facebook.com/intern/diff/D42049013)

tugsbayasgalan committed:
TEMP

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "TEMP"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "TEMP"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

zhxchen17 commented on torch/csrc/autograd/python_variable_indexing.h:
```diff
@@ -1,11 +1,91 @@
+};
+
+// This mirrors Cpython's PySlice_Unpack method
+inline UnpackedSlice __PySlice_Unpack(PyObject* _r) {
```
Is inlining this function intentional? I think it makes more sense to put definition into cpp file.

voznesenskym commented:
Lots of test failures :(

voznesenskym commented on torch/csrc/autograd/python_variable_indexing.h:
```diff
@@ -1,11 +1,91 @@
+
+// This mirrors Cpython's PySlice_Unpack method
+inline UnpackedSlice __PySlice_Unpack(PyObject* _r) {
+  PySliceObject* r = (PySliceObject*)_r;
```
Why did this change files? This makes it harder to review...

voznesenskym commented:
If this needs ATen / core native changes, change that first + add tests + try to minimize code motion there. Then do the dynamo change on top of that.

voznesenskym reviewed (changes requested):
If this needs ATen / core native changes, change that first + add tests + try to minimize code motion there. Then do the dynamo change on top of that.

albanD commented on aten/src/ATen/TensorIndexing.h:
```diff
@@ -39,50 +39,65 @@ struct TORCH_API Slice final {
     }
     if (!start_index.has_value()) {
-      start_ = step_ < 0 ? INDEX_MAX : 0;
+      TORCH_CHECK_VALUE(!step_.is_symbolic(), "slice step cannot be symbolic");
```
If you expect this here, then you don't need to handle the case above?

tugsbayasgalan replied:
updated

albanD commented on aten/src/ATen/TensorIndexing.h:
```diff
@@ -213,14 +228,14 @@ namespace impl {
     const c10::optional<SymIntArrayRef>& self_sizes) {
   // TODO: implement negative step
-  TORCH_CHECK_VALUE(step > 0, "step must be greater than zero");
+  TORCH_CHECK_VALUE(step.expect_int() > 0, "step must be greater than zero");
```
No need for expect here either.

tugsbayasgalan replied:
updated

albanD commented on torch/csrc/autograd/python_variable_indexing.cpp:
```diff
@@ -133,14 +133,8 @@ inline Variable valueToTensor(
-  if (!THPUtils_unpackSlice(index, start_ptr, stop_ptr, step_ptr)) {
-    throw python_error();
-  }
+static inline UnpackedSlice checkUnpackSlice(PyObject* index) {
```
nit: this indirection is not needed anymore.

tugsbayasgalan committed:
Update on "Support DynamicSlicing in torchdynamo"


Currently, i am limiting the step size to be normal integer, just because I am not really sure what the expected behaviour for x[None:None:symint] as the original python behaviour adjusts start and stop based on step. 


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42049013).

tugsbayasgalan committed:
Update on "Support DynamicSlicing in torchdynamo"


Currently, i am limiting the step size to be normal integer, just because I am not really sure what the expected behaviour for x[None:None:symint] as the original python behaviour adjusts start and stop based on step. 


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

Differential Revision: [D42049013](https://our.internmc.facebook.com/intern/diff/D42049013)

[ghstack-poisoned]

tugsbayasgalan commented on aten/src/ATen/TensorIndexing.h:
```diff
@@ -39,50 +39,65 @@ struct TORCH_API Slice final {
-      if (step_ < -INDEX_MAX)
-        step_ = -INDEX_MAX;
+      if (!step_.is_symbolic()) {
+        TORCH_CHECK_VALUE(step_.expect_int() != 0, "slice step cannot be zero");
```
updated

tugsbayasgalan commented on aten/src/ATen/TensorIndexing.h:
```diff
@@ -39,50 +39,65 @@ struct TORCH_API Slice final {
-      // does "step = -step" as part of a slice reversal.
-      if (step_ < -INDEX_MAX)
-        step_ = -INDEX_MAX;
+      if (!step_.is_symbolic()) {
```
updated

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42049013).

tugsbayasgalan committed:
Update on "Support DynamicSlicing in torchdynamo"


Currently, i am limiting the step size to be normal integer, just because I am not really sure what the expected behaviour for x[None:None:symint] as the original python behaviour adjusts start and stop based on step. 


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

Differential Revision: [D42049013](https://our.internmc.facebook.com/intern/diff/D42049013)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42049013).

tugsbayasgalan committed:
Update on "Support DynamicSlicing in torchdynamo"


Currently, i am limiting the step size to be normal integer, just because I am not really sure what the expected behaviour for x[None:None:symint] as the original python behaviour adjusts start and stop based on step. 


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

Differential Revision: [D42049013](https://our.internmc.facebook.com/intern/diff/D42049013)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42049013).

tugsbayasgalan committed:
Update on "Support DynamicSlicing in torchdynamo"


Currently, i am limiting the step size to be normal integer, just because I am not really sure what the expected behaviour for x[None:None:symint] as the original python behaviour adjusts start and stop based on step. 


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

Differential Revision: [D42049013](https://our.internmc.facebook.com/intern/diff/D42049013)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support DynamicSlicing in torchdynamo"


Currently, i am limiting the step size to be normal integer, just because I am not really sure what the expected behaviour for x[None:None:symint] as the original python behaviour adjusts start and stop based on step. 


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

Differential Revision: [D42049013](https://our.internmc.facebook.com/intern/diff/D42049013)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42049013).

ezyang commented:
This is obsoleted by https://github.com/pytorch/pytorch/pull/91341 right?
